### ISSUE TITLE
sigil_bs4's prettyprint_xhtml doubles xml declaration

### ISSUE THREAD

**Issue by wrCisco** [Sigil-Ebook/Sigil#123]
Hi, great work with the latest release! Keep it up!
I was experimenting with the new sigil_bs4 module (on Mac OS X) and I found that prettyprint_xhtml doesn't check if there is already an xml declaration at the beginning of the file that is being changed.
If I write a plugin like this:
```python
from sigil_bs4 import _soup
def run(bk):
    for idFile, hrefFile in bk.text_iter():
        soup = _soup(bk.readfile(idFile))
        bk.writefile(idFile, soup.prettyprint_xhtml())
    return 0
```
all xhtml files in my epub begin with two <?xml version="1.0" encoding="utf-8"?>. Am I missing something? As a temporary workaround I subclassed sigil_bs4.BeautifulSoup and overwrote prettyprint_xhtml method to make the check (something like this:
```python
if super(sigil_bs4.BeautifulSoup,
         self).prettyprint_xhtml(indent_level, eventual_encoding,
               formatter, indent_chars).startswith("<?xml"):
            return super(sigil_bs4.BeautifulSoup,
                         self).prettyprint_xhtml(indent_level, 
                               eventual_encoding, formatter, indent_chars)
else:
    return prefix + super(sigil_bs4.BeautifulSoup,
                          self).prettyprint_xhtml(indent_level, 
                                eventual_encoding, formatter, indent_chars)
```
. Surely you can write something better).

On the other hand, if I import sigil_gumbo_bs4_adapter and call prettyprint_xhtml on the output of the function parse, it adds the second xml declaration as a comment after the DOCTYPE (no harm done), but divides autoconclusive tags like `<link>` in a starttag and an endtag.

**Comment by kevinhendricks** [comment]
Hi,
Yes the internal GumboInterface inside of Sigil always removes the xml header before processing it with gumbo and then adds it back.  gumbo itself does not want to see a xml header at all (it is really an html5 parser).

As for sigil_bs4, yes the xml header is an issue.  It should be removed before feeding the file into prettyprint if you are parsing with html5lib or gumbo in the background.

I will look into fixing both of those issues hopefully before our next release.  We are just having weird issues on Windows 7 that is consuming all of my time trying to figure things out.

Thanks for your report.

Kevin

**Comment by kevinhendricks** [comment]
I have thought about this some more.  To know how to handle this, I need to know the backend parser that you tell BS4 to use:

The possible values are: 
    "lxml-xml" - tells BS4 to use the lxml xml parser, (alternate name: "xml")
    "html.parser" - tells BS4 to use the html parser that is built into both python 2.7 or 3.4
    "html5lib" - tells BS4 to use the html5lib parser
    "lxml-html" - tells BS4 to use the lxml html parser

Each handles xml declaration headers in a different way.

Please note that "lxml-xml", "html5lib", and "lxml-html" are always available with the embedded Python 3.4 built into Sigil.  If you want to use these parsers via python 2.7, you must instruct the user how to install these extra libraries into their external python 2.7 interpreter.

Our sigil_bs4 version of BS4 can also be used as a wrapper around Google's gumbo parser by importing the sigil_gumbo_bs4_adapter.py.  Please note, gumbo can not accept xml header declarations as these are not valid in html5 parsers (it makes them comments).  Our python adapter to gumbo accepts utf-8 or full uncode text as input.  Any full unicode text is utf-8 encoded automatically before being passed to gumbo as gumbo only accepts utf-8 encoded text.   Also, I have now modified the gumbo adapter code to remove any xml declaration before processing to prevent extraneous comments from being added.

Second thing is that I have now modified decodexml(), prettyprint_xhtml() and serialize_xhtml() in sigil_bs4 to strip off any existing xml declaration header that is found after serializing the code, and replace it with a new one that properly sets the intended final encoding of the file.

So hopefully with this information and these changes you should not have to worry about double xml header declarations anymore as along as you serialize all xml via decodexml() and all xhtml via either serialize_xhtml() or prettyprint_xhtml() all of which only exist in our version of BS4.  Users we decide to use official BS4 interfaces like decode() or prettyprint() will have to live with the exact same output as would be generated by official bs4 (except for the lxml related namespace fixes which I hope to eventually upstream).

Hope this helps.

Look for these changes in the next release.

If you are looking for a fast workaround, try the following before sending the text to the parser:

import re
def _remove_xml_header(data):
    return re.sub(r'<\s*\?xml\s*[^\?>]*\?*>\s*','',data, flags=re.I)

text = _remove_xml_header(text)

Thanks for your bug report I am closing this now.  If the fix does not for any reason appear in the next release, please re-open the issue.

Thanks for your bug report!

Kevin

**Comment by wrCisco** [comment]
Ok, thanks! I'll try that.
(By the way, currently I use html.parser for consistency reasons with the usage in older versions of Sigil - I used that to make things as simple as possible for non-programmer users and have nonetheless predictable output).
